topdown/strings.go

```diff
@@ -410,38 +422,160 @@ func builtinTrimSpace(a ast.Value) (ast.Value, error) {
 }
 
 func builtinSprintf(a, b ast.Value) (ast.Value, error) {
+	var resultString ast.String
 	s, err := builtins.StringOperand(a, 1)
+
 	if err != nil {
 		return nil, err
 	}
 
 	astArr, ok := b.(*ast.Array)
+
 	if !ok {
 		return nil, builtins.NewOperandTypeErr(2, b, "array")
 	}
 
 	args := make([]interface{}, astArr.Len())
 
-	for i := range args {
-		switch v := astArr.Elem(i).Value.(type) {
-		case ast.Number:
-			if n, ok := v.Int(); ok {
-				args[i] = n
-			} else if b, ok := new(big.Int).SetString(v.String(), 10); ok {
-				args[i] = b
-			} else if f, ok := v.Float64(); ok {
-				args[i] = f
-			} else {
-				args[i] = v.String()
+	var formatStringArr []string
+	var explicitExprArr []string
+	var indexes []formatIndexMap
+
+	if len(args) > 0 {
+		formatStringArr = formatRE.FindAllString(s.String(), -1)
+		explicitExprArr = detectExpArgsRE.FindAllString(s.String(), -1)
+	}
+
+	if len(explicitExprArr) > 0 { // Parse the explicit indexes
+		// Iterate through the format string arr and
+		// explicit expression array and find their respective indexes.
+		for fIndex, formatString := range formatStringArr {
+			for _, explicitExpStr := range explicitExprArr {
+				if formatString == explicitExpStr {
+					// Grab the index from the format string
+					// remove the brackets from the integer
+					// and the argument index (args), and store it in an
+					// array of struct (formatIndexMap)
+					elementIndexStr := regexp.MustCompile(`\[\d\]`).FindAllString(explicitExpStr, -1)[0]
+					elementIndexStr = elementIndexStr[1 : len(elementIndexStr)-1]
+					elementIndex, _ := strconv.Atoi(elementIndexStr)
+					index := formatIndexMap{
+						fIndex, elementIndex - 1,
+					}
+					indexes = append(indexes, index)
+					break
+				}
 			}
-		case ast.String:
-			args[i] = string(v)
-		default:
-			args[i] = astArr.Elem(i).String()
 		}
+
+		var modifiedArgs []interface{}
+		argsIndex := 0
+
+		// Parse the explicit indexes
+		for i, formatString := range formatStringArr {
+			isFormatted := false
+			for _, index := range indexes {
+				if index.formatArrIndex == i {
+					// Trim the parent format string (s) by removing the quotes
+					// and remove all of the explicit indexes from (s)
+					trimmedParentFmtString := s.String()[1 : len(s.String())-1]
+					formattedString := regexp.MustCompile(`\[\d\]`).ReplaceAllString(trimmedParentFmtString, "")
+					s = ast.String(formattedString)
+
+					switch v := astArr.Elem(index.arrayIndex).Value.(type) {
+					case ast.Number:
+						// Format the numbers
+						if strings.Contains(formatString, "f") {
+							float, _ := v.Float64()
+							modifiedArgs = append(modifiedArgs, float)
+						} else {
+							if in, ok := v.Int(); ok {
+								modifiedArgs = append(modifiedArgs, in)
+							} else if bin, ok := new(big.Int).SetString(v.String(), 10); ok {
+								modifiedArgs = append(modifiedArgs, bin)
+							} else {
+								modifiedArgs = append(modifiedArgs, v.String())
+							}
```

Comment:
Isn't this block used a few times? I wonder if we could spin it out into a helper method... 🤔

Reply:
Let me work a little magic here.

Reply:
Done. Committing latest changes now with the helper method.